glib/src/win32.rs

```diff
@@ -1,16 +1,103 @@
 // Take a look at the license at the top of the repository in the LICENSE file.
 
-use std::path::PathBuf;
+use std::{ffi::OsString, path::PathBuf};
 
 use crate::translate::*;
 
-#[doc(alias = "g_win32_get_package_installation_directory_of_module")]
-pub fn win32_get_package_installation_directory_of_module(
-    hmodule: ffi::gpointer,
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub enum OSType {
+    #[doc(alias = "G_WIN32_OS_ANY")]
+    Any,
+    #[doc(alias = "G_WIN32_OS_WORKSTATION")]
+    Workstation,
+    #[doc(alias = "G_WIN32_OS_SERVER")]
+    Server,
+}
+
+#[doc(hidden)]
+impl IntoGlib for OSType {
+    type GlibType = ffi::GWin32OSType;
+
+    #[inline]
+    fn into_glib(self) -> Self::GlibType {
+        match self {
+            Self::Any => ffi::G_WIN32_OS_ANY,
+            Self::Workstation => ffi::G_WIN32_OS_WORKSTATION,
+            Self::Server => ffi::G_WIN32_OS_SERVER,
+        }
+    }
+}
+
+#[doc(alias = "g_win32_check_windows_version")]
+pub fn check_windows_version(major: i32, minor: i32, spver: i32, os_type: OSType) -> bool {
+    unsafe {
+        from_glib(ffi::g_win32_check_windows_version(
+            major,
+            minor,
+            spver,
+            os_type.into_glib(),
+        ))
+    }
+}
+
+#[doc(alias = "g_win32_get_command_line")]
+#[doc(alias = "get_command_line")]
+pub fn command_line() -> Vec<OsString> {
+    unsafe { FromGlibPtrContainer::from_glib_full(ffi::g_win32_get_command_line()) }
+}
+
+#[doc(alias = "g_win32_error_message")]
+pub fn error_message(error: i32) -> OsString {
+    unsafe { from_glib_full(ffi::g_win32_error_message(error)) }
+}
+
+#[doc(alias = "g_win32_getlocale")]
+pub fn getlocale() -> OsString {
+    unsafe { from_glib_full(ffi::g_win32_getlocale()) }
+}
+
+#[doc(alias = "g_win32_get_package_installation_directory")]
+#[doc(alias = "get_package_installation_directory")]
+pub fn package_installation_directory(
+    package: Option<&str>,
+    dll_name: Option<&str>,
 ) -> Option<PathBuf> {
+    unsafe {
+        from_glib_full(ffi::g_win32_get_package_installation_directory(
+            package.to_glib_none().0,
+            dll_name.to_glib_none().0,
+        ))
+    }
+}
+
+#[doc(alias = "g_win32_get_package_installation_directory_of_module")]
+#[doc(alias = "get_package_installation_directory_of_module")]
+pub fn package_installation_directory_of_module(hmodule: ffi::gpointer) -> Option<PathBuf> {
```

Comment:
> gtk4-win32 already depends on windows-rs

The dependency is optional btw

Reply:
After testing the GLib wrapper function with bogus pointers generated using `rand::random::<usize>() as std::os::windows::raw::HANDLE`, I can attest that the function just runs gracefully.

Though, there's something interesting going on with the underlying `GetModuleFilenameW` function: if it failed to infer the DLL referenced by the pointer, Win32's `GetLastError` function will return `WIN32_ERROR(126)`, a tuple containing a `pub u32` that can be passed into `g_win32_error_message`. 

Is it allowed to make the safe wrapper return `Result<PathBuf, WIN32_ERROR>` instead so that the code can be marked as safe without compromising behavioral predictability?

Reply:
That seems like a good plan